CA: Add GenerateCRL gRPC method

aarongable opened the pull request:
Add a new CA gRPC method named `GenerateCRL`. In the
style of the existing `GenerateOCSP` method, this new endpoint
is implemented as a separate service, for which the CA binary
spins up an additional gRPC service.

This method uses gRPC streaming for both its input and output.
For input, the stream must contain exactly one metadata message
identifying the crl number, issuer, and timestamp, and then any
number of messages identifying a single certificate which should
be included in the CRL. For output, it simply streams chunks of
bytes.

Fixes #6161

jsha commented on ca/crl.go:
```diff
@@ -0,0 +1,164 @@
+	}
+
+	if lifetime == 0 {
+		logger.Warningf("got zero for crl lifetime; setting to default 9 days")
```
Why offer a default here rather than error out?

jsha commented on ca/crl.go:
```diff
@@ -0,0 +1,164 @@
+
+func (ci *crlImpl) GenerateCRL(stream capb.CRLGenerator_GenerateCRLServer) error {
+	rcs := make([]pkix.RevokedCertificate, 0)
+	var number *big.Int
+	var thisUpdate time.Time
+	var issuer *issuance.Issuer
```
It was a little surprising to see these declarations hanging up so far from their use. What if we just store a `*capb.GenerateCRLRequest_Metadata` object outside the loop, and do the processing (getting issuer, making bigint, etc) once the loop exits? That reduces our ability to early-exit for unknown issuers, but it shortens the loop and lets us reduce predeclaration. This would also let us use a nil check to figure out "did we already get metadata?"

We could also do much of the `capb.GenerateCRLRequest_Entry` branch in its own function that produces a pkix.RevokedCertificate.

jsha commented on ca/proto/ca.proto:
```diff
@@ -50,3 +50,24 @@ message GenerateOCSPRequest {
+message GenerateCRLRequest {
+  oneof payload {
+    CRLMetadata metadata = 1;
+    core.CRLEntry entry = 2;
```
Why is CRLEntry in `core`?

jsha commented on ca/crl.go:
```diff
@@ -0,0 +1,164 @@
+		NextUpdate:          thisUpdate.Add(-time.Second).Add(ci.lifetime),
+	}
+
+	crlBytes, err := x509.CreateRevocationList(
```
This needs an audit log.

For OCSP, we log every certificate we generate OCSP for, plus its status (we should add reasonCode). We chunk these up into about 38 per log line, IIRC.

For CRLs, I think we should do something similar, though we can omit the certificates that are not listed in the CRL. So, once we have the list of RevokedCertificates, we should log N lines, containing M certificates per line, with their status and reasonCode. We should then log the final revocation list creation event, with the important parameters (ThisUpdate, NextUpdate, Number, length of RevokedCertificates, maybe summary stats of reasonCode?), a hash of the final CRL, and its size. It would be nice to be able to log the hash of the tbsCertList but I don't think Go's current APIs give us that.

jsha replied:
Let's consider this good for now, and file a followup issue to plumb through the reason code once #6198 is done.

jsha reviewed (changes requested):
This generally looks great! I was a bit surprised by the fact that we stream all the inputs and stream all the outputs, but in the middle they are both fully buffered in memory. It looks like that's that only option we've got, given the APIs we have, though.

I was briefly confused about the CRLNumber extension being generated from ThisUpdate. That should be documented somewhere. Maybe on the .proto?

The new `crl.go` is fairly independent of the `ca` package. If practical, it should be its own package.

Some RFC 5280 stuff:

> Conforming CRL issuers MUST NOT use CRLNumber
>   values longer than 20 octets.

Does Go take check this for us? If not, we should document how we satisfy this MUST (by generating CRLNumber from a timestamp that is < 20 octets).

> If a CRL issuer generates two CRLs (two complete CRLs, two delta
>   CRLs, or a complete CRL and a delta CRL) for the same scope at
>   different times, the two CRLs MUST NOT have the same CRL number.
>   That is, if the this update field ([Section 5.1.2.4](https://datatracker.ietf.org/doc/html/rfc5280#section-5.1.2.4)) in the two CRLs
>   are not identical, the CRL numbers MUST be different.

I believe we satisfy this by having different scopes for different CRLs with the same ThisUpdate, right? In other words, the caller is responsibly for satisfying this requirement?

I documented some more requirements in #6188.

aarongable commented:
> This generally looks great! I was a bit surprised by the fact that we stream all the inputs and stream all the outputs, but in the middle they are both fully buffered in memory. It looks like that's that only option we've got, given the APIs we have, though.

Yep. I really wanted to be able to do a streaming hash as the bytes come in, but the crypto/x509 package doesn't give us an option for that. If this becomes a performance bottleneck we could of course write our own signature routine from scratch, but I'd obviously prefer to avoid that if at all possible.

> I was briefly confused about the CRLNumber extension being generated from ThisUpdate. That should be documented somewhere. Maybe on the .proto?

Good call. I did this at J.C.'s suggestion, as a good way to ensure that all the CRLNumbers of a set of shards line up, and as a way to guarantee that the CRLNumber is constantly increasing without having to store extra state elsewhere. I've added this note to the comment on the proto message.

> The new `crl.go` is fairly independent of the `ca` package. If practical, it should be its own package.

I did this to explicitly parallel `//ca/ocsp.go`, which has the same structure and the same inclusion in the larger CA service. Should we move that into another package as well? If yes, can I do both as a follow-up to this PR?

> Some RFC 5280 stuff:
> 
> > Conforming CRL issuers MUST NOT use CRLNumber
> > values longer than 20 octets.
> 
> Does Go take check this for us? If not, we should document how we satisfy this MUST (by generating CRLNumber from a timestamp that is < 20 octets).

Ooh, I'd totally forgotten this requirement! It looks like Go [does not enforce](https://cs.opensource.google/go/go/+/refs/tags/go1.18.3:src/crypto/x509/x509.go;l=2129;drc=abb3f055246cfda3ca7d4cd2179c636ce207c265) this. I've [filed a bug](https://github.com/golang/go/issues/53543) to have them do so, but also documented that our timestamps will never be longer than 20 octets.

> 
> > If a CRL issuer generates two CRLs (two complete CRLs, two delta
> > CRLs, or a complete CRL and a delta CRL) for the same scope at
> > different times, the two CRLs MUST NOT have the same CRL number.
> > That is, if the this update field ([Section 5.1.2.4](https://datatracker.ietf.org/doc/html/rfc5280#section-5.1.2.4)) in the two CRLs
> > are not identical, the CRL numbers MUST be different.
> 
> I believe we satisfy this by having different scopes for different CRLs with the same ThisUpdate, right? In other words, the caller is responsibly for satisfying this requirement?

Right. The idea behind CRL Shards is "each of these has a different scope, even if it's really hard to articulate exactly what that scope is". Functionally, this doesn't matter at all, as there is no language that I'm aware of which prevents two scopes from overlapping, and therefore containing most (or even all) of the same revoked certs.

> I documented some more requirements in #6188.

Thanks! I'll tackle that next week.

jsha commented:
> I did this to explicitly parallel //ca/ocsp.go, which has the same structure and the same inclusion in the larger CA service. Should we move that into another package as well? If yes, can I do both as a follow-up to this PR?

Yes to both.

beautifulentropy commented on core/proto/core.proto:
```diff
@@ -93,3 +93,10 @@ message Order {
+  int32 reason = 2;
+  int64 revokedAt = 3; // Unix timestamp (nanoseconds)
+}
+ 
```
```suggestion
 
 
```

aarongable commented:
Okay, with the exception of the suggested forking+refactoring discussed above (which I'm looking into now), I have addressed all comments and consider this PR to be ready for review.

I've filed https://github.com/letsencrypt/boulder/issues/6196 for the "move this into it's own package" follow-up, and we already have https://github.com/letsencrypt/boulder/issues/6188 for more linting/safety follow-up.

jsha reviewed (approved):
Looks good!

beautifulentropy commented on cmd/boulder-ca/main.go:
```diff
@@ -56,10 +59,16 @@ type Config struct {
 
-		// LifespanOCSP is how long OCSP responses are valid for; It should be longer
-		// than the minTimeToExpiry field for the OCSP Updater.
+		// LifespanOCSP is how long OCSP responses are valid for. It should be
```
Love the addition of this comment; I took a look at the logic in main where this is loaded from config and the constructor to which the parsed duration is passed I don't see anywhere where we validate this.

beautifulentropy commented on ca/crl.go:
```diff
@@ -0,0 +1,221 @@
+	}
+
+	if lifetime == 0 {
+		logger.Warningf("got zero for crl lifetime; setting to default 9 days")
+		lifetime = 9 * 24 * time.Hour
+	} else if lifetime >= 10*24*time.Hour {
+		return nil, fmt.Errorf("crl lifetime cannot be more than 10 days, got: %s", lifetime)
+	} else if lifetime <= 0*time.Hour {
+		return nil, fmt.Errorf("crl lifetime must be positive, got: %s", lifetime)
```
To make these messages more useful to an operator you could provide, in quotes, the (implicit) name of this field in the JSON config.

Similarly, quoting the parsed input can be helpful as well.

aarongable replied:
Unfortunately the code here inside the constructor has no knowledge of the config keys used to set these arguments -- I think "crl lifetime" is close enough to "crlLifetime" to be figured out?

I've quoted the values.

beautifulentropy replied:
Works for me!

beautifulentropy commented on ca/crl.go:
```diff
@@ -0,0 +1,221 @@
+			if !ok {
+				return fmt.Errorf("got unrecognized IssuerNameID: %d", payload.Metadata.IssuerNameID)
+			}
+
```
```suggestion
```

aarongable replied:
My personal preference is to have a blank line after every `if err != nil {...}` or `if !ok {...}`. It serves to break the code into nice logical chunks: do as much setup as you like, then see if it worked, then break and move on to the next operation.

beautifulentropy commented on ca/crl.go:
```diff
@@ -0,0 +1,221 @@
+			if err != nil {
+				return err
+			}
+
```
```suggestion
```

beautifulentropy reviewed (changes requested):
Small nits, otherwise everything looks great!
